### PR TITLE
Cypress e2e Model Serving Tests ODS-2276 and ODS-2279

### PR DESCRIPTION
**Pull request by antowaddle**
## Description
https://issues.redhat.com/browse/RHOAIENG-12651
Migrate Robot Tests ODS-2276 & ODS-2279 to Cypress. 

## How Has This Been Tested?
- An oc login should be performed in the cluster before running the test
- test-variables.yml should be configured properly
- Export the path to the test-variables.yml: $ export CY_TEST_CONFIG=<path_to>/test-variables.yml
- Test has been tested against ODH-Nightly cluster and [PSI-11](https://issues.redhat.com//browse/PSI-11)

image
image

Headless
Go to odh-dashboard/frontend/src/tests/cypress and run the below commands

npx cypress run --spec "cypress/tests/e2e/settings/servingRuntimes/testMultiModelServingRuntimeCreation.cy.ts" --browser chrome
npx cypress run --spec "cypress/tests/e2e/settings/servingRuntimes/testSingleServingRuntimeCreation.cy.ts" --browser chrome 

## Test Impact
None - this is a test.

## Request review criteria:

Self checklist (all need to be checked):
- [ ] The developer has manually tested the changes and verified that the changes work
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 

- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

File: frontend/src/__tests__/cypress/cypress/tests/e2e/settings/servingRuntimes/testMultiModelServingRuntimeCreation.cy.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    // Navigate to serving runtimes and import a multi-model serving yaml
+    cy.step('Navigate to Serving Runtimes and add a new serving runtime').then(() => {
+      // TODO: Remove extended timeout once '/servingruntimes' performance is optimized
+      // Current workaround for ODH page loading performance issues
+      return cy.wrap(servingRuntimes.navigate(), { timeout: 100000 });
+    });
+
+    cy.log('Navigation completed, looking for Add button');
+    cy.then(() => {
```
Comment (emilys314) [inline frontend/src/__tests__/cypress/cypress/tests/e2e/settings/servingRuntimes/testMultiModelServingRuntimeCreation.cy.ts:44]: What is the purpose of using `cy.then(() ...)` for `servingRuntimes.findAddButton()` ? 

I saw there were 2 other references of `findAddButton()` that did not need this

File: frontend/src/__tests__/cypress/cypress/tests/e2e/settings/servingRuntimes/testMultiModelServingRuntimeCreation.cy.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      return servingRuntimes.findAddButton();
+    })
+      .should('exist')
+      .should('exist', { timeout: 100000 })
```
Comment (emilys314) [inline frontend/src/__tests__/cypress/cypress/tests/e2e/settings/servingRuntimes/testMultiModelServingRuntimeCreation.cy.ts:48]: Do we only need one of the `.should('exist')`?

File: frontend/src/__tests__/cypress/cypress/pages/servingRuntimes.ts [hunk 3]
```diff
@@ -123,6 +123,22 @@ class ServingRuntimes {
...
+
+  deleteModel() {
+    return cy.findByRole('menuitem', { name: 'Delete' });
+  }
+
+  deleteModal() {
+    return cy.get('[data-testid="delete-modal-input"]');
+  }
+
+  deleteModelServingButton() {
```
Comment (emilys314) [inline frontend/src/__tests__/cypress/cypress/pages/servingRuntimes.ts:139]: Based on the guidelines in `docs/testing.md` for naming functions, they should start with `find...`

**Review by emilys314 — COMMENTED** [review]
There is also a failing cypress test (that one seem quite flaky) so

File: frontend/src/__tests__/cypress/cypress/fixtures/resources/files/modelServing/multiModel/ovms_servingruntime.yaml
Comment (manosnoam) [inline frontend/src/__tests__/cypress/cypress/fixtures/resources/files/modelServing/multiModel/ovms_servingruntime.yaml:1]: Can you place this file in the yaml resources ?
`main/frontend/src/__tests__/cypress/cypress/fixtures/resources/yaml/modelServing`

or even better - without subfolder "yaml":
`main/frontend/src/__tests__/cypress/cypress/fixtures/resources/modelServing`

File: frontend/src/__tests__/cypress/cypress/pages/servingRuntimes.ts (hunk 3 above)
Comment (manosnoam) [reply frontend/src/__tests__/cypress/cypress/pages/servingRuntimes.ts:139]: agree, it just returns the locator, but does not click on it yet.

File: frontend/src/__tests__/cypress/cypress/tests/e2e/settings/servingRuntimes/testMultiModelServingRuntimeCreation.cy.ts (hunk 2 above)
Comment (antowaddle) [reply frontend/src/__tests__/cypress/cypress/tests/e2e/settings/servingRuntimes/testMultiModelServingRuntimeCreation.cy.ts:48]: Good catch, resolved now.

File: frontend/src/__tests__/cypress/cypress/tests/e2e/settings/servingRuntimes/testMultiModelServingRuntimeCreation.cy.ts (hunk 1 above)
Comment (antowaddle) [reply frontend/src/__tests__/cypress/cypress/tests/e2e/settings/servingRuntimes/testMultiModelServingRuntimeCreation.cy.ts:44]: I was facing some performance issues when executing the tests on ODH vs. RHOAI. Initially I had to chain the commands in this way for the tests to run successfully. I've reviewed this again this morning and I've got it working with the cy.then, I'll commit that now. 

Thanks for all your comments, really appreciate it.

File: frontend/src/__tests__/cypress/cypress/fixtures/resources/files/modelServing/multiModel/ovms_servingruntime.yaml
Comment (antowaddle) [reply frontend/src/__tests__/cypress/cypress/fixtures/resources/files/modelServing/multiModel/ovms_servingruntime.yaml:1]: Sure, I've now removed the 'files' directory and moved these files into resources.

File: frontend/src/__tests__/cypress/cypress/pages/servingRuntimes.ts [hunk 4]
```diff
@@ -123,6 +123,22 @@ class ServingRuntimes {
   getRowById(id: string) {
     return new ServingRuntimeRow(id);
   }
+
+  findEditModel() {
+    return cy.findByLabelText('Kebab toggle');
```
Comment (FedeAlonso) [inline frontend/src/__tests__/cypress/cypress/pages/servingRuntimes.ts:128]: The "Kebab togle" returns the Edit functionality?

File: frontend/src/__tests__/cypress/cypress/tests/e2e/settings/servingRuntimes/testMultiModelServingRuntimeCreation.cy.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      .within(() => {
+        servingRuntimes.findEditModel().click();
+        servingRuntimes.findDeleteModel().click();
+      });
+
+    servingRuntimes.findDeleteModal().should('be.visible').type(metadataDisplayName);
+
+    cy.step(`Delete the model ${modelServingName}`);
+    servingRuntimes.findDeleteModelServingButton().click();
+    servingRuntimes.getRowById(modelServingName).find().should('not.exist');
```
Comment (FedeAlonso) [inline frontend/src/__tests__/cypress/cypress/tests/e2e/settings/servingRuntimes/testMultiModelServingRuntimeCreation.cy.ts:78]: Could we add some backend validation?

**Review by FedeAlonso — CHANGES_REQUESTED** [review]

File: frontend/src/__tests__/cypress/cypress/pages/servingRuntimes.ts (hunk 4 above)
Comment (antowaddle) [reply frontend/src/__tests__/cypress/cypress/pages/servingRuntimes.ts:128]: Yes, that's what the element is called at present.

**Pull request merged**